cli/lepton-cli.1.in
.TH lepton-cli 1 "@DATE@" "Lepton EDA" @VERSION@
.SH NAME
lepton-cli - Lepton EDA Command-Line Utility
.SH "SYNOPSIS"
.B lepton-cli
[\fIOPTION\fR ...] \fICOMMAND\fR [\fIARGS\fR ...]
.SH "DESCRIPTION"
.PP
.B lepton-cli
is part of the Lepton EDA (Electronic Design Automation) toolset.
It provides a number of small command-line utilities for working
with schematic and symbol files, and is designed to be used for
batch processing of designs created using the schematic editor
\fBlepton-schematic\fR(1).  It currently has three built-in
\fICOMMAND\fRs:

.B lepton-cli export
is used to create SVG, PDF, PNG, PS and EPS files from schematic and
symbol files, for printing or embedding in other documents.

.B lepton-cli config
allows reading and writing settings in Lepton EDA project, user and system
configuration stores.

.B lepton-cli shell
provides a Scheme REPL for command-line batch processing of schematic
data.

.SH "GENERAL OPTIONS"
.TP 8
\fB--no-rcfiles\fR
Prevent `gafrc' Scheme initialisation files from being loaded.
.TP 8
\fB-h\fR, \fB--help\fR
Print a help message.
.TP 8
\fB-V\fR, \fB--version\fR
Print \fBlepton-cli\fR version information.

.SH "EXPORTING IMAGE FILES"
.B lepton-cli export
[\fIOPTION\fR ...] \fB-o\fR \fIOUTPUT\fR [\fI--\fR] \fIFILE\fR ...

.B lepton-cli export
can export schematic and symbol files in a variety of image formats
for printing or further processing.  It currently supports single-page
PNG, SVG and EPS output, and multi-page PS and PDF output.  It accepts
a variety of options for controlling how the output is formatted.

.TP 8
\fB-o\fR, \fB--output\fR=\fIFILE\fR
Output generated image data to \fIFILE\fR.
.TP 8
\fB-f\fR, \fB--format\fR=(\fBpng\fR | \fBpdf\fR | \fBsvg\fR | \fBps\fR | \fBeps\fR)
Specify an output format.  Usually, this option is not required,
because \fBlepton-cli export\fR will infer the correct format from
the file extension of the output file.
.TP 8
\fB-p\fR, \fB--paper\fR=\fINAME\fR
Size the output for a particular paper size.  The \fINAME\fR should be
a PWG 5101.1-2002 paper name.  For example, valid values include
`iso_a4', `na_letter', or `na_d'.
.TP 8
\fB-P\fR, \fB--paper-names\fR
Print paper size names suitable to pass to the \fB--paper\fR option
and exit.
.TP 8
\fB-l\fR, \fB--layout\fR=(\fBauto\fR | \fBlandscape\fR | \fBportrait\fR)
When using a paper size, set the orientation of the output.  If `auto'
layout is used, the orientation that best fits the drawing will be
used.
.TP 8
\fB-s\fR, \fB--size\fR=(\fBauto\fR | \fIWIDTH\fR:\fIHEIGHT\fR)
Size the output with specific dimensions.  If the size is `auto',
select the size that best fits the drawing.
.TP 8
\fB-k\fR, \fB--scale\fR=\fIFACTOR\fR Set the output scale
\fIFACTOR\fR. This is a distance identical with 100 points (1
default grid spacing) in \fBlepton-schematic\fR(1) coordinate
space. It is used to size the output when neither \fB--paper\fR
nor \fB--size\fR are given, and defaults to 100 mil.
.TP 8
\fB-m\fR, \fB--margins\fR=(\fBauto\fR | \fITOP\fR[:\fILEFT\fR[:\fIBOTTOM\fR[:\fIRIGHT\fR]]])
Set the widths of the margins to be used.  If `auto' margins are
specified, a sensible default value will be chosen.  Up to four margin
widths can be provided.  If one is provided, it will be used on all
four sides.  If two are provided, the first will be used for the
top/bottom and the second for the left/right.  If three are provided,
the first will be used for the top, the second for left/right, and the
third for the bottom.
.TP 8
\fB-a\fR, \fB--align\fR=(\fBauto\fR | \fIHALIGN\fR:\fIVALIGN\fR)
Set how the drawing is aligned within the page.  \fIHALIGN\fR controls
the horizontal alignment, and \fIVALIGN\fR the vertical.  Each
alignment value should be in the range 0.0 to 1.0.  The `auto'
alignment is equivalent to a value of `0.5:0.5', i.e. centered.
.TP 8
\fB-d\fR, \fB--dpi\fR=\fIDPI\fR
Set the number of pixels per inch used when generating PNG output.
.TP 8
\fB-c\fR, \fB--color\fR
Enable colour output.
.TP 8
\fB--no-color\fR
Disable color output.
.TP 8
\fB-F\fR, \fB--font\fR=\fIFONT-FAMILY\fR
Set the font to be used for drawing text.
.TP 8
\fB--\fR
Treat all remaining arguments as schematic or symbol filenames.  Use
this if you have a schematic or symbol filename which begins with `-'.

.PP
The \fB--size\fR, \fB--margins\fR, or \fB--scale\fR options described above accept
values using units of `mm', `cm', `in', `pc', `px', or `pt'.  If you
do not provide a unit, points are assumed.  N.b. that `px' are
evaluated relative to the current \fB--dpi\fR setting.

.PP
When using the \fB--size\fR, \fB--margins\fR, or \fB--align\fR options
with multiple values, you may use `;', or ` ' as a separator between
them instead of `:'. In such a case, remember to properly quote your
arguments to avoid them to be interpreted by your shell.


.SH "ACCESSING CONFIGURATION"
.B lepton-cli config
[\fIOPTION\fR] [\fIGROUP\fR \fIKEY\fR [\fIVALUE\fR]]

.B lepton-cli config
is a basic utility for inspecting and modifying Lepton EDA configuration
stores.

.PP
If a \fIGROUP\fR and \fRKEY\fR are specified, retrieves the value of
that configuration parameter.  If a \fIVALUE\fR is specified, sets
the value of that parameter.  The \fB-p\fR, \fB-u\fR and \fB-s\fR
options can be used to select the configuration store affected (by
default, the project configuration store for the current working
directory).  If no \fIGROUP\fR and \fIKEY\fR are provided, outputs
the filename of the selected configuration store.

.PP
.TP 8
\fB-p\fR, \fB--project\fR[=\fIPATH\fR]
Select the project configuration store of the project located in
\fIPATH\fR.  If no \fIPATH\fR is provided, selects the project in the
current working directory.
.TP 8
\fB-u\fR, \fB--user\fR
Select the user configuration store.
.TP 8
\fB-s\fR, \fB--system\fR
Select the system configuration store.  Depending on user permissions,
the system configuration store may be read-only.
.TP 8
\fB-c\fR, \fB--cache\fR
Select the program-specific configuration store (CACHE configuration context,
associated with $XDG_CACHE_HOME/lepton-eda/gui.conf configuration file).

.SH "SCHEME PROCESSING"
.B lepton-cli shell
[\fIOPTION ...\fR]

.B lepton-cli shell
provides a Scheme Read-Eval-Print Loop (REPL) for automating
processing of schematic and symbol files.  It is designed to be used
with the Lepton EDA Scheme API.

.TP 8
\fB-s\fR \fIFILE\fR
Evaluate Scheme source code from \fIFILE\fR, and exit.
.TP 8
\fB-c\fR \fIEXPR\fR
Evaluate the Scheme expression \fIEXPR\fR, and exit.
.TP 8
\fB--\fR
Stop scanning arguments; run interactively.
.TP 8
\fB-L\fR \fIDIRECTORY\fR
Prepend \fIDIRECTORY\fR to the list of directories to be searched for
Scheme files.
.TP 8
\fB-l\fR \fIFILE\fR
Evaluate Scheme source code from \fIFILE\fR.

.PP
The \fB-s\fR, \fB-c\fR and \fB--\fR switches stop argument processing
and pass all the remaining arguments as the value of `(command-line)'.

.SH AUTHORS
See the `AUTHORS' file included with this program.

.SH COPYRIGHT
.nf
Copyright \(co 2012-2017 gEDA Contributors.
Copyright \(co 2017-@YEAR@ Lepton Developers.
License GPLv2+: GNU GPL version 2 or later. Please see the `COPYING'
file included with this program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
\fBlepton-schematic\fR(1).

For more information on the Scheme API, see the \fBlepton-scheme\fR
Texinfo manual.  If the \fBinfo\fR program is properly installed at
your site, the command

.IP
.B
info lepton-scheme

.PP
should give you access to the complete manual.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2016 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 1.9.3
====================================

* The gEDA/gaf source tarball no longer includes libintl.  When
  building gEDA/gaf with internationalisation support, a GNU
  gettext-compatible libintl is now required.

* gEDA/gaf core tools will now automatically rotate per-user log
  files.  250 log files per tool will be retained from the last 24
  hours, and 50 log files per tool from all time.

* gEDA/gaf tools now search for data and configuration files in the
  standard XDG directories.  By default, these are:

  - `/usr/local/share/gEDA/` and `/usr/share/gEDA` for system data
  - `/etc/xdg/gEDA/` for system configuration
  - `$HOME/.local/share/gEDA/` for per-user data
  - `$HOME/.config/gEDA/` for per-user configuration.

  However, gEDA/gaf tools will still use `$HOME/.gEDA` for per-user
  files if that directory exists, they still obey the `$GEDADATA` and
  `$GEDADATARC` environment variables, and by default the search path
  includes the installation prefix selected when compiling gEDA/gaf.

* The configure script now supports an `--enable-relocatable` switch.
  When specified, gEDA/gaf tools do not have the configured prefix
  compiled into them.

* The configure script now supports a `--disable-deprecated` switch.
  When specified, various deprecated features and behaviour are
  disabled, which may break backward-compatibility with some
  configurations.

* A new Scheme API function, `log!`, has been added to the `(geda
  log)` module.  It allows Scheme code to emit log messages in the
  same way that the tools' C code does.

* Gschem multiattrib dialog improvements:

  - Shift+Tab can be used to move focus backward from `value' to
    `name'.

  - Tooltip to the `value' text entry has been added explaning how
    to enter newline or tab.

  - Horizontal scrollbar for `value' has been added to prevent
    resizing of the dialog.

* Gschem undo information now contains center of viewport and
  current page scale, so if gschem window size changes before
  undo, page will be zoomed accordingly.

* Faster redrawing of pages in gschem.

* Menu entry to launch Scheme REPL in gschem's background terminal
  has been added.

* Changed calculation of the size of bus and net junctions.

* `Bom` gnetlist backend has been refactored, all its functions
  have help now.

* View bounds in gschem window has been changed so symbols can be
  edited without translation.

* New Scheme module for part list creation has been added for
  gnetlist.

* Netlist output of several gnetlist backends is sorted by package
  names.

* New Scheme procedures for working with source library have been
  added to the `(geda library)` module: source-library-contents
  and set-source-library-contents!.  Please see the "geda-scheme"
  info manual for more details.


Notable changes in gEDA/gaf 1.9.2
=================================
* Build system changes:

  - Guile 2.0 or newer is now required.

  - top-level control file allowing continuous integration using
    Travis CI has been added.

* Nullor expansion has been fixed in the `spice' and `spice-sdb'
  backends.

  - Element name prefix "E_" has been substituted for "E-" to prevent
    errors output by some simulators (e.g. ngspice).

  - A default gain value has been added in the nullor expansion code.

* Two obsolete scripts verilog2vhdl.sh and script.sed have been
  removed.

* Undocumented and ambiguous feature of the `spice-sdb' backend for
  gnetlist consisting in the possibility of using the "value="
  attribute instead of "file=" only for include blocks and only in
  embedding mode has been removed.

* The following scripts and programs have been moved from the `utils'
  directory to a new `contrib' directory and are no longer installed
  by default:

         smash_megafile          olib
         gmk_sym                 convert_sym
         sarlacc_schem           sarlacc_sym
         pads_backannotate       sw2asc
         gnet_hier_verilog.sh

* `gpstoimage' has been removed from the distribution; equivalent
  functionality is provided by the `gaf' command.

* `convert_sym.awk' has been removed from the distribution; equivalent
  functionality is provided by the `convert_sym' command.

* gnetlist now supports output of netlists to the standard output.

* New Scheme procedures and hooks have been added.  Please see the
  "geda-scheme" info manual for more details.

  - `set-component-with-transform!' is a modified version of
    `set-component!', which applies all the specified transformations
    to a component and immediately updates it.

  - `auto-refdes-reset!' resets components' refdeses

  - `copy-objects-hook' is called whenever an object is copied, it can
    be used, for example, to reset component's refdes on copy.

* The following gschem configuration items have been added:

  - `third-button-cancel' cancels draw actions by the third mouse
    button in mousepan mode.

  - `grid-mode' sets default grid mode.

* Fixes/updates to the *BSD, Mac OS X, and MinGW builds.

* Changes in the gschem UI event state machine:

  - New drawing modes apply immediately.

  - Indication of an action being in progress has been added.

  - Mirror action can now be used during an object placement.

  - Show/hide text function is now enabled during other actions.

  - RMB panning is now enabled during placing or drawing.

* Some gschem dialog have been merged, made non-modal, and placed in
  sidebar:

  - Color, line, fill and pin type dialogs have been merged into one
    non-modal object properties dialog with widgets applying values
    immediately.

  - Text properties dialog.


cli/lepton-cli.c
/*
 * Lepton EDA command-line utility
 * Copyright (C) 2012-2013 Peter Brett <peter@peter-b.co.uk>
 * Copyright (C) 2012-2014 gEDA Contributors
 * Copyright (C) 2017-2020 Lepton EDA Contributors
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <config.h>
#include <version.h>

#include <locale.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <getopt.h>

/* Gettext translation */
#include "gettext.h"

#include <glib.h>

#include "builtins.h"

#include <liblepton/liblepton.h>


#define short_options "+hV"

static struct option long_options[] =
  {
    {"help", 0, NULL, 'h'},
    {"version", 0, NULL, 'V'},
    {"no-rcfiles", 0, NULL, 2},
    {NULL, 0, NULL, 0},
  };

struct internal_command {
  const char *name;
  int (*func)(int, char **);
};

static struct internal_command commands[] =
  {
    {"shell", cmd_shell},
    {"config", cmd_config},
    {"export", cmd_export},
    {NULL, NULL},
  };

/* Print help info and exit */
static void
usage (void)
{
  printf (_("Usage: lepton-cli [OPTION...] COMMAND [ARGS ...]\n"
"\n"
"Lepton EDA command-line utility.\n"
"\n"
"General options:\n"
"  --no-rcfiles   inhibit loading of 'gafrc' files\n"
"  -h, --help     display usage information and exit\n"
"  -V, --version  display version information and exit\n"
"\n"
"Commonly-used commands (type `lepton-cli <cmd> --help' for usage):\n"
"  shell          Scheme REPL for interactive Lepton EDA data processing\n"
"  config         Edit Lepton EDA configuration\n"
"  export         Export Lepton EDA files in various image formats.\n"
"\n"
"Report bugs at <%1$s>\n"
"Lepton EDA homepage: <%2$s>\n"),
    PACKAGE_BUGREPORT,
    PACKAGE_URL);

  exit (0);
}

/* Print version info and exit */
static void
version (void)
{
  char* msg = version_message();
  printf ("%s\n", msg);
  free (msg);

  exit (0);
}

int
main (int argc, char **argv)
{
  int c;
  char *cmd = NULL;
  int cmd_argc = 0;
  char **cmd_argv = NULL;
  int (*cmd_func)(int, char **) = NULL;

  /* Set up gettext */
#if ENABLE_NLS
  setlocale (LC_ALL, "");
  bindtextdomain ("lepton-cli", LOCALEDIR);
  textdomain ("lepton-cli");
  bind_textdomain_codeset ("lepton-cli", "UTF-8");
#endif

  while (-1 != (c = getopt_long (argc, argv, short_options,
                                 long_options, NULL))) {
    switch (c) {

    case 0:
      /* This is a long-form-only flag option, and has already been
       * dealt with by getopt_long(). */
      break;

    case 2: /* --no-rcfiles */
      g_setenv ("LEPTON_INHIBIT_RC_FILES", "1", 1);
      break;

    case 'V':
      version ();

    case 'h':
      usage ();

    case '?':
      /* getopt_long already printed an error message */
      fprintf (stderr, _("\nRun `lepton-cli --help' for more information.\n"));
      exit (1);
      break;

    default:
      g_assert_not_reached ();
    }
  }

  /* The next argument should be a command */
  if (optind == argc) {
    fprintf (stderr,
             _("ERROR: You must specify a command to run.\n"
               "\n"
               "Run `lepton-cli --help' for more information.\n"));
    exit (1);
  }

  cmd = argv[optind];

  /* Look up the command */
  int i;
  for (i = 0; commands[i].name != NULL; i++) {
    if (strcmp (cmd, commands[i].name) == 0) {
      cmd_func = commands[i].func;
      break;
    }
  }
  if (cmd_func == NULL) {
    fprintf (stderr,
             _("ERROR: Unrecognised command `%1$s'.\n"
               "\n"
               "Run `lepton-cli --help' for more information.\n"),
             cmd);
    exit (1);
  }

  cmd_argc = argc - optind;
  cmd_argv = argv + optind;
  optind = 1;

  return cmd_func (cmd_argc, cmd_argv);
}
